### ISSUE TITLE
[BUG] Randomly failed to create volume with backing image

### ISSUE THREAD

**Issue by yangchiu** [longhorn/longhorn#7543]
## Describe the bug (🐛 if you encounter this issue)

Test case `test_csi_backup_with_backing_image` randomly fails when creating volume with backing image with error message:
```
longhorn.ApiError: (ApiError(...), '500 : failed to create volume: unable to create volume longhorn-testvol-wa52q0: 
admission webhook "mutator.longhorn.io" denied the request: backup volume\'s backing image  does not match backup\'s backing image bi-test. 
The backup volume resource is being updated and please retry later\n{\'code\': 500, \'detail\': \'\', \'message\': \'failed to create volume: unable to create volume longhorn-testvol-wa52q0: 
admission webhook "mutator.longhorn.io" denied the request: backup volume\\\'s backing image  does not match backup\\\'s backing image bi-test. 
The backup volume resource is being updated and please retry later\', \'status\': 500}')
```

https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5844/testReport/tests/test_backing_image/test_csi_backup_with_backing_image_nfs_1_10_/

Not sure it's a bug or there is any restriction? 

## To Reproduce

Run test case `test_csi_backup_with_backing_image`

## Environment

 - Longhorn version: master-head
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl):
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version:
   - Number of management node in the cluster:
   - Number of worker node in the cluster:
 - Node config
   - OS type and version:
   - Kernel version:
   - CPU per node:
   - Memory per node:
   - Disk type(e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes:
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):
 - Number of Longhorn volumes in the cluster:
 - Impacted Longhorn resources:
   - Volume names:

**Comment by ChanYiLin** [comment]
It is related to "Backup Backing Image" feature.
It seems a rare case 1/10 happened.

The root cause is 
- somehow a "BACKINGIMAGE" got created as a BackVolume in the backupstore
- But there is no such volume, and the name is also not valid to create the BackupVolume CR, so the BackupTargetController failed to pull it
- at the same time, test case started to create the test volume with its own BackupVolume, but it hasn't synced yet because BackupTargetController kept failing.

I am now investigating why there is a "BACKINGIMAGE" BackupVolume in the backupstore.

**Comment by innobead** [comment]
Is it possible that there were some leftover resource in the previous tests not cleaned up to cause this? Could check tests in longhorn-tests

**Comment by ChanYiLin** [comment]
"BACKINGIMAGE" backup volume is found in backupstore means 
- Longhorn create a  BackupVolume config in backupstore which name is "BACKINGIMAGE"
- so when BackupTargetController tried to list the BackupVolume from the backupstore
- it got this config and the name is "BACKINGIMAGE"
- so it then tried to create the BackupVolume CRD for it but failed because CR name should not have uppercase.

That is, no matter what, there should not have BackupVolume config in backupstore with name "BACKINGIMAGE"

**Comment by ChanYiLin** [comment]
Oh I think I know the reason
It is because of the lock
Our lock mechanism will create the lock file in the backup volume directory and the lock name is the same as backup volume
I use the lock name "BACKINGIMAGE" for all the backup backing image
And since there is no such backup volume, so when I require the lock in BackingImage backup, it then create the directory "BACKINGIMAGE" in the backupstore under the backup volume directory
This make BackupTargetController confused and think there is a backup volume needs to be pulled whose name is "BACKINGIMAGE"
(because we get the backup volume name by parsing the path...)

**Comment by innobead** [comment]
> I use the lock name "BACKINGIMAGE" for all the backup backing image And since there is no such backup volume, so when I require the lock in BackingImage backup, it then create the directory "BACKINGIMAGE" in the backupstore under the backup volume directory This make BackupTargetController confused and think there is a backup volume needs to be pulled whose name is "BACKINGIMAGE" (because we get the backup volume name by parsing the path...)

Does this we can't do back or restore the backing image in parallel?

**Comment by ChanYiLin** [comment]
No the locking system wont blick the parallel backup
It is designed in our locking mechanism 
https://github.com/longhorn/longhorn/blob/master/enhancements/20200701-backupstore-file-locks.md#proposal

**Comment by chriscchien** [comment]
Test [result](https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5872/testReport/tests/test_backing_image/)

**Comment by yangchiu** [comment]
@ChanYiLin I can still find this error on master-head (longhorn-manager 7a56d7):

https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5898/
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5897/
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5893/
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5891/
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5890/

Is it possible that there is another issue here?

**Comment by ChanYiLin** [comment]
@yangchiu I will take a look, thanks

**Comment by ChanYiLin** [comment]
I think this is kind of timing issue,
The info of BackupVolume is not yet synced from backupstore, so the field `BackingImage` is empty
which leads to the error
```
backup volume\'s backing image  does not match backup\'s backing image bi-test. The backup volume resource is being updated and please retry later
``` 

I will try to add a time delay and a check before restoring the next volume

**Comment by ChanYiLin** [comment]
Confirmed the issue, it is because of the timing issue in the tests
should wait until BackupVolume CR is synced before restore the Volume
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5904/console
```
test_backing_image.py::test_csi_backup_with_backing_image[nfs-1-10] [DEBUG] find bv: {'backingImageChecksum': '', 'backingImageName': '', 'created': '', 'dataStored': '', 'labels': None, 'lastBackupAt': '', 'lastBackupName': '', 'messages': None, 'name': 'longhorn-testvol-h8x2g3', 'size': '', 'storageClassName': ''}
FAILED
test_backing_image.py::test_csi_backup_with_backing_image[nfs-2-10] [DEBUG] find bv: {'backingImageChecksum': 'bd79ab9e6d45abf4f3f0adf552a868074dd235c4698ce7258d521160e0ad79ffe555b94e7d4007add6e1a25f4526885eb25c53ce38f7d344dd4925b9f2cb5d3b', 'backingImageName': 'bi-test', 'created': '2024-01-09T11:37:21Z', 'dataStored': '16777216', 'labels': {'KubernetesStatus': '{"pvName":"longhorn-testvol-w2w5fe","pvStatus":"Bound","namespace":"default","pvcName":"longhorn-testvol-w2w5fe","lastPVCRefAt":"","workloadsStatus":[{"podName":"csi-backup-test","podStatus":"Running","workloadName":"","workloadType":""}],"lastPodRefAt":""}', 'VolumeRecurringJobInfo': '{}', 'longhorn.io/volume-access-mode': 'rwo'}, 'lastBackupAt': '2024-01-09T11:37:21Z', 'lastBackupName': 'backup-db89cd0fdbec43e2', 'messages': {}, 'name': 'longhorn-testvol-w2w5fe', 'size': '3221225472', 'storageClassName': 'longhorn-csi-pv-test'}
[DEBUG] find bv: {'backingImageChecksum': '304f3ed30ca6878e9056ee6f1b02b328239f0d0c2c1272840998212f9734b196371560b3b939037e4f4c2884ce457c2cbc9f0621f4f5d1ca983983c8cdf8cd9a', 'backingImageName': 'bi-test', 'created': '2024-01-09T11:40:07Z', 'dataStored': '16777216', 'labels': {'KubernetesStatus': '{"pvName":"longhorn-testvol-0dogqc","pvStatus":"Bound","namespace":"default","pvcName":"longhorn-testvol-0dogqc","lastPVCRefAt":"","workloadsStatus":[{"podName":"csi-backup-test","podStatus":"Running","workloadName":"","workloadType":""}],"lastPodRefAt":""}', 'VolumeRecurringJobInfo': '{}', 'longhorn.io/volume-access-mode': 'rwo'}, 'lastBackupAt': '2024-01-09T11:40:07Z', 'lastBackupName': 'backup-8b3b7ab0837e4dfe', 'messages': {}, 'name': 'longhorn-testvol-0dogqc', 'size': '3221225472', 'storageClassName': 'longhorn-csi-pv-test'}
PASSED
```

**Comment by ChanYiLin** [comment]
Hi @yangchiu 
https://github.com/longhorn/longhorn-tests/pull/1668 PR is merged
Please help to retest the issue, thanks!

**Comment by yangchiu** [comment]
Verified passed on master-head (longhorn-tests a8db3c6). Run test case `test_csi_backup_with_backing_image` for 100 times, all passed.

Test results: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5933/ ~ https://ci.longhorn.io/job/private/job/longhorn-tests-regression/5942/
